Miscellaneous fixes...

nphilipp opened the pull request:
...what I said on standup that day wasn't just empty promises. It just keeps on growing. :wink:

* Vagrant: fix how the IPA server is provisioned
* Tests: fix one unittest that fails on Python 3.8+, configure Tox to use Python 3.8, too
* Fix import order
* Don't use 3rd party `mock` package (there is `unittest.mock` and we use it in most places already)
* Improve a test which is unnecessarily strict, breaking when unrelated things change

nphilipp committed:
Make "vagrant provision" work on unclean installs

Previously, you'd have to destroy the whole box. Now it checks if it
finds the installation log file of the IPA server, and if so, uninstalls
it before installing it freshly.

Signed-off-by: Nils Philippsen <nils@redhat.com>

nphilipp committed:
Tests: Fix how flask.session is patched

From Python 3.8 on, unittest.mock examines target objects to
differentiate between sync and async callables. Flasks' session,
request, etc. proxies don't tolerate being inspected outside of a
request context well and upstream has no intention of accommodating
this, see: https://github.com/pallets/flask/issues/3454

Instead, patch the session object with a new MagicMock object which
doesn't know or care about the thing it replaces.

Signed-off-by: Nils Philippsen <nils@redhat.com>

nphilipp committed:
Tox: Run tests on Python 3.8

Signed-off-by: Nils Philippsen <nils@redhat.com>

ryanlerch reviewed (approved):
LGTM! Just the single conflict to resolve, but other than that. +1

nphilipp committed:
Set the umask expected by `ipa-server-install`

Otherwise it may fail when running the playbook.

Signed-off-by: Nils Philippsen <nils@redhat.com>

nphilipp committed:
Use unittest.mock consistently

We require Python >= 3.6 which ships mock as part of the unittest module
(which we use in most places anyway), therefore it's unnecessary to pull
in the 3rd party mock package which fills in the gap for Python versions
before 3.3.

Signed-off-by: Nils Philippsen <nils@redhat.com>

nphilipp committed:
Don't ignore flake8 errors/warnings unnecessarily

Signed-off-by: Nils Philippsen <nils@redhat.com>

nphilipp committed:
Fix import order and specify conventions

Additionally, require flake8-import-order for testing.

Signed-off-by: Nils Philippsen <nils@redhat.com>

nphilipp committed:
Make test more robust against unrelated changes

Only verify the parts of the URL we care about, so changes in secrets
etc. won't make the test fail.

Signed-off-by: Nils Philippsen <nils@redhat.com>

abompard commented on devel/create-test-data.py:
```diff
@@ -1,8 +1,8 @@
 
-import python_freeipa
 from faker import Faker
+import python_freeipa
```
Are you sure? In [PEP8](https://www.python.org/dev/peps/pep-0008/#id23) they sort imports with `import x` first and `from x import y` second.

In the [Google style guide](https://google.github.io/styleguide/pyguide.html?showone=Imports_formatting#313-imports-formatting) you can see both sortings, and in the [example file](https://github.com/PyCQA/flake8-import-order/blob/master/tests/test_cases/complete_google.py) that flake8-import-order provides the `import x` come before the `from x import y`.

nphilipp replied:
I've changed `import-order-style` to `smarkets` in tox.ini which doesn't require 3rd party packages to be separate from each other, but places `from ... import ...` beneath plain `import ...` statements within their respective blocks.

abompard commented on tox.ini:
```diff
@@ -74,8 +74,11 @@ commands =
 exclude = .git,.tox,dist,*egg
 
+# Configure flake8-import-order
+application-import-names = noggin, noggin_messages
```
You need to add `flake8-import-order` to the dev dependencies if you want to use this feature.

nphilipp replied:
Done.
